File: android/core/src/main/kotlin/com/tickernest/core/HoldingUpsert.kt
package com.tickernest.core

/**
 * Pure-domain holding upsert planner — Kotlin twin of the Node module.
 * Lets the Android client simulate the diff locally so the edit dialog can
 * show "Update / Reduce / Full Exit" instantly without a server round trip.
 *
 * The server is still authoritative; this is purely for UI affordances.
 */

data class HoldingState(val qty: Money, val avgCost: Money)

data class PlannedSoldShare(
    val qty: Money,
    val costBasisAtSell: Money,
    val soldPrice: Money?,
    val reason: String?,
    val mistake: String?,
)

sealed interface Plan {
    data object Noop : Plan
    data class Insert(val next: HoldingState) : Plan
    data class Update(val next: HoldingState, val soldShare: PlannedSoldShare?) : Plan
    data class FullExit(val soldShare: PlannedSoldShare) : Plan
}

data class HoldingUpsertRequest(
    val desired: HoldingState,
    val soldPrice: Money? = null,
    val reason: String? = null,
    val mistake: String? = null,
)

fun planHoldingUpsert(current: HoldingState?, req: HoldingUpsertRequest): Plan {
    val desired = req.desired
    if (current == null) {
        return if (desired.qty.signum() <= 0) Plan.Noop
               else Plan.Insert(desired)
    }
    val sameQty = current.qty.compareTo(desired.qty) == 0
    val sameAvg = current.avgCost.compareTo(desired.avgCost) == 0
    if (sameQty && sameAvg) return Plan.Noop

    if (desired.qty < current.qty) {
        val soldQty = current.qty.subtract(desired.qty)
        val ss = PlannedSoldShare(
            qty = soldQty,
            costBasisAtSell = current.avgCost,  // OLD avg, frozen
            soldPrice = req.soldPrice,
            reason = req.reason,
            mistake = req.mistake,
        )
        return if (desired.qty.isZero()) Plan.FullExit(ss)
               else Plan.Update(desired, ss)
    }
    return Plan.Update(desired, soldShare = null)
}

File: services/api/src/holding/holding.upsert.ts
/**
 * Pure-domain holding upsert planner.
 *
 * Avg-cost is user-supplied; we never derive it from a trade history.
 * The one rule we DO enforce: when an edit decreases qty, we synthesise a
 * SoldShare row with `cost_basis_at_sell` = the AVG AT EDIT TIME (the old avg).
 * This snapshot is what makes the Sold-Shares journal historical even when the
 * user later changes their avg.
 *
 * This module is pure: given (current, requested), it returns a Plan that the
 * persistence layer applies inside a single TX.
 */
import { D, Money } from '../common/types/money';

export interface HoldingState {
  qty: Money;
  avgCost: Money;
}

export interface HoldingUpsertRequest {
  /** What the user wants the row to look like AFTER the write. */
  desired: HoldingState;
  /** Optional sell-side metadata if the user is recording a sell explicitly. */
  soldPrice?: Money;
  reason?: string;
  mistake?: string;
}

export type Plan =
  | { kind: 'NOOP' }
  | { kind: 'INSERT'; next: HoldingState }
  | { kind: 'UPDATE'; next: HoldingState; soldShare?: PlannedSoldShare }
  | { kind: 'DELETE'; soldShare: PlannedSoldShare };

export interface PlannedSoldShare {
  qty: Money;
  costBasisAtSell: Money; // snapshot of old avg
  soldPrice: Money | null;
  reason: string | null;
  mistake: string | null;
}

export const planHoldingUpsert = (
  current: HoldingState | null,
  req: HoldingUpsertRequest,
): Plan => {
  const { desired } = req;

  // Brand new holding
  if (!current) {
    if (desired.qty.lte(0)) {
      // Inserting a "holding" of qty=0 makes no sense; reject upstream.
      return { kind: 'NOOP' };
    }
    return { kind: 'INSERT', next: desired };
  }

  const sameQty = current.qty.eq(desired.qty);
  const sameAvg = current.avgCost.eq(desired.avgCost);
  if (sameQty && sameAvg) return { kind: 'NOOP' };

  // Qty decreased → record a SoldShare (cost basis snapshotted from OLD avg).
  if (desired.qty.lt(current.qty)) {
    const soldQty = current.qty.sub(desired.qty);
    const soldShare: PlannedSoldShare = {
      qty: soldQty,
      costBasisAtSell: current.avgCost, // immutable historical snapshot
      soldPrice: req.soldPrice ?? null,
      reason: req.reason ?? null,
      mistake: req.mistake ?? null,
    };
    if (desired.qty.isZero()) {
      return { kind: 'DELETE', soldShare };
    }
    return { kind: 'UPDATE', next: desired, soldShare };
  }

  // Qty same or increased → straight update, no SoldShare.
  return { kind: 'UPDATE', next: desired };
};
